### PR TITLE
Search suggestions by static attribute

### LINKED ISSUE #5874: `from_rows` and `new` suggestions in CB while opening with the `Table` selected
### Discord username

_No response_

### What type of issue is this?

Permanent – Ocurring repeatably

### Is this issue blocking you from using Enso?

- [ ] Yes, I can't use Enso because of this issue.

### Is this a regression?

- [ ] Yes, previous version of Enso did not have this issue.

### What issue are you facing?

When I open a CB with `Table` selected I am getting `from_rows` and `Table.new` as the best suggestions - they are invalid and choosing those would create always a node with an error.
image

### Expected behaviour

We should not suggest both of those when working with already created table.

### How we can reproduce it?

_No response_

### Screenshots or screencasts

_No response_

### Enso Version

Nightly 9.03.2023

### Browser or standalone distribution

Standalone distribution

### Browser Version or standalone distribution

standalone

### Operating System

MacOS

### Operating System Version

_No response_

### Hardware you are using

_No response_

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #5874 

Changelog:
- add: `isStatic` parameter to `search/completion` request to search by the `static` suggestion attribute
- update: search non-static suggestions when opening component browser

### Important Notes

Component browser doesn't show `Table.new` and `Table.from_rows` suggestions when a `Table` node is selected.

2023-03-21-151117_1301x877_scrot

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: app/gui/src/controller/searcher.rs [hunk 1]
```diff
@@ -1218,7 +1218,7 @@ impl Searcher {
             let requests = return_types_for_engine.into_iter().map(|return_type| {
                 info!("Requesting suggestions for returnType {return_type:?}.");
                 let file = graph.module.path().file_path();
-                ls.completion(file, &position, &this_type, &return_type, &tags)
+                ls.completion(file, &position, &this_type, &return_type, &tags, &Some(false))
```
Comment (farmaazon) [inline app/gui/src/controller/searcher.rs]: But now we always filter out static methods, don't we? We should do this only if we have this argument specified.

We should filter out static methods only then `this_type` has a type:
```rust
let this_type = this_type.await;
let is_static = this_type.is_some().then_some(false).
info!("Requesting new suggestion list. Type of `self` is {this_type:?}.");
...
    ls.completion(file, &position, &this_type, &return_type, &tags, &is_static)
```

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui/src/controller/searcher.rs [hunk 2]
```diff
@@ -1218,7 +1218,8 @@ impl Searcher {
             let requests = return_types_for_engine.into_iter().map(|return_type| {
                 info!("Requesting suggestions for returnType {return_type:?}.");
                 let file = graph.module.path().file_path();
-                ls.completion(file, &position, &this_type, &return_type, &tags)
+                let is_static = if this_type.is_some() { Some(false) } else { None };
```
Comment (farmaazon) [inline app/gui/src/controller/searcher.rs]: Please put it outside the closure (as I suggested) - there is no need to recompute it for every return_type.

Also, the `if` may be replaced with `then_some` method: https://doc.rust-lang.org/std/primitive.bool.html#method.then_some

File: app/gui/controller/engine-protocol/src/language_server.rs [hunk 3]
```diff
@@ -177,6 +177,7 @@ trait API {
     , self_type   : Option<String>
     , return_type : Option<String>
     , tags        : Option<Vec<SuggestionEntryType>>
+    , is_static   : Option<bool>
```
Comment (JaroslavTulach) [inline app/gui/controller/engine-protocol/src/language_server.rs:180]: Shouldn't the attribute have a default value?

File: docs/language-server/protocol-language-server.md [hunk 4]
```diff
@@ -4368,6 +4368,8 @@ Sent from client to the server to receive the autocomplete suggestion.
   returnType?: string;
   // Filter by the suggestion types
   tags?: [SuggestionEntryType];
+  // Filter by `static` attribute of the suggestion
+  isStatic?: Boolean;
```
Comment (JaroslavTulach) [inline docs/language-server/protocol-language-server.md:4372]: What is the default value when the attribute is missing?

File: engine/language-server/src/main/scala/org/enso/languageserver/search/SearchApi.scala [hunk 5]
```diff
@@ -76,7 +76,8 @@ object SearchApi {
       position: Position,
       selfType: Option[String],
       returnType: Option[String],
-      tags: Option[Seq[SuggestionKind]]
+      tags: Option[Seq[SuggestionKind]],
+      isStatic: Option[Boolean]
```
Comment (JaroslavTulach) [inline engine/language-server/src/main/scala/org/enso/languageserver/search/SearchApi.scala:80]: Can one ask for both - static/instance - methods at once? Or one needs two queries?

Or is it never needed to ask for both types at once?

**Review by JaroslavTulach — APPROVED** [review]
Please describe what is the default value of the `isStatic` attribute - otherwise OK.

File: app/gui/controller/engine-protocol/src/language_server.rs (hunk 3 above)
Comment (4e6) [reply app/gui/controller/engine-protocol/src/language_server.rs:180]: These are the parameters of a search request. If you don't want to specify a parameter, you pass None

File: docs/language-server/protocol-language-server.md (hunk 4 above)
Comment (4e6) [reply docs/language-server/protocol-language-server.md:4372]: the TypeScript `?` notation in the type definition, i.e.
```ts
{
    a?: Type;
}
```
means that the `a` attribute is optional and may be `null`

File: engine/language-server/src/main/scala/org/enso/languageserver/search/SearchApi.scala (hunk 5 above)
Comment (4e6) [reply engine/language-server/src/main/scala/org/enso/languageserver/search/SearchApi.scala:80]: > Can one ask for both - static/instance - methods at once?

If you don't care about the `static` attribute, you just pass `None`. This way the method will return both static and non-static suggestions.

**Review by farmaazon — APPROVED** [review]

File: lib/scala/searcher/src/test/scala/org/enso/searcher/sql/SuggestionsRepoTest.scala [hunk 6]
```diff
@@ -1698,6 +1723,32 @@ class SuggestionsRepoTest extends AnyWordSpec with Matchers with RetrySpec {
...
+          _   <- repo.insert(suggestion.function)
+          _   <- repo.insert(suggestion.local)
+          res <-
+            repo.search(
+              None,
+              Seq(),
+              None,
+              None,
+              None,
+              Some(true)
```
Comment (hubertp) [inline lib/scala/searcher/src/test/scala/org/enso/searcher/sql/SuggestionsRepoTest.scala:1743]: For completeness, would be probably good to have a test that checks non-static methods only

**Review by hubertp — APPROVED** [review]

**Pull request merged**